File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *temp = head;
        int cnt = 0;
        while (temp != NULL) {
            cnt++;
            temp = temp->next;
        }
        int nod = cnt - n + 1;
        if (head == NULL) return head;
        if (nod == 1) {
            ListNode *temp = head;
            head = head->next;
            delete temp;
            return head;
        }
        int chk = 0;
        temp = head;
        ListNode *prev = nullptr;
        while (temp != NULL) {
            chk++;
            if (chk == nod) {
                prev->next = temp->next;
                delete temp;
                break;
            }
            prev = temp;
            temp = temp->next;
        }
        return head;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # so we are using two poiter and gap algo for this it means fast move till none and maintain n gaps so when we reach it then slow next is candidate

        dummy = ListNode(next=head) # in case of n = 1
        fast = dummy
        slow= dummy
        for _ in range(n):
            fast = fast.next
        while(fast.next !=None):
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return dummy.next